lib/datadog/statsd/uds_connection.rb

```diff
@@ -14,18 +14,26 @@ def initialize(socket_path, **kwargs)
         super(**kwargs)
 
         @socket_path = socket_path
+        @socket = nil
+        connect
+      end
+
+      def close
+        @socket.close if @socket
+        @socket = nil
       end
 
       private
 
       def connect
-        socket = Socket.new(Socket::AF_UNIX, Socket::SOCK_DGRAM)
-        socket.connect(Socket.pack_sockaddr_un(@socket_path))
-        socket
+        close unless @socket == nil
+
+        @socket = Socket.new(Socket::AF_UNIX, Socket::SOCK_DGRAM)
+        @socket.connect(Socket.pack_sockaddr_un(@socket_path))
       end
 
       def send_message(message)
-        socket.sendmsg_nonblock(message)
+        @socket.sendmsg_nonblock(message)
       rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::ENOENT => e
         @socket = nil
```

Comment:
Should this use `close` as well? Just occurred to me that this may lead us to leak sockets

Reply:
Good question, I expected the socket to not be opened if we received these exceptions, but I'll check this out, especially if `#close` is a noop if already closed. 👀

Reply:
Actually, I think we should not close here and this is wrong to `@socket = nil` here.
We want the caller of `#send_message` to take care of this (in my case, the retry mechanism in `connection.rb`)